docs/design/sparse-v2.md

```diff
@@ -0,0 +1,225 @@
+- `jj sparse set --add exclude:foo/bar` adds a new `DirExclusion` type rule
+- `jj sparse set --exclude foo/bar` as a possible shorthand for the above
+- `jj sparse list` will print the explicit rules
+
+Paths will be stored in an ordered, canonical form which unambiguously describes
+the set of files to be included. Every `--add` command will append to the end of
+this list before the patterns are canonicalized. Whether a file is included is
+determined by the first matching rule in reverse order.
+
+For example:
+
+```txt
+include:foo
+exclude:foo/bar
+include:foo/bar/baz
+exclude:foo/bar/baz/qux
+```
+
+Produces rule set which includes "foo/file.txt", excludes "foo/bar/file.txt",
+includes "foo/bar/baz/file.txt", and excludes "foo/bar/baz/qux/file.txt". If the
+rules are subtly re-ordered, they become canonicalized to a smaller, but
+functionally equivalent form:
+
+```txt
+# Before
+include:foo
+exclude:foo/bar/baz/qux
+include:foo/bar/baz
+exclude:foo/bar
+
+# Canonicalized
+include:foo
+exclude:foo/bar
+```
+
+## Working Copy Map
+
+WARNING: This section is intentionally lacking, more research is needed.
+
+All Sparse Patterns v2 will come equipped with a default no-op mapping.
+
+```rust
+vec![WorkingCopyMap {
+    src_path: RepoPathBuf::root(),
+    dst_path: RepoPathBuf::root(),
+}]
+```
+
+`SparsePatterns` will provide an interface to map client paths into repo paths,
+and vice versa. The WorkingCopy trait will apply this mapping to all snapshot
+and checkout operations, and and jj commands which accept relative paths will
+need to be updated to perform client path -> repo path translations as needed.
+It's not clear at this time _which_ commands will need changing, as some are
+more likely to refer to repo paths rather than client paths.
+
+TODO: Expand this section.
+
+In particular, the path rules for sparse patterns for _always_ be repo paths,
+not client paths. Thus, if the client wants to track "foo" and rename it to
+"subdir/bar", they must `jj sparse set --add foo` and
+`jj client-map set --from foo --to bar`. In other words, the mapping operation can
+be thought of as always _after_ the sparse operation.
+
+New commands will enable editing of the WorkingCopyMap:
+
+- `jj client-map list` will print all mapping pairs.
+- `jj client-map set --from foo --to bar` will add a new mapping to the end of the list.
+- `jj client-map remove --from foo` will remove a specific mapping rule.
+- `jj client-map edit` will pull up a text editor for manual editing.
+
+Like sparse paths, mapping rules are defined to apply in _order_ and on any
+save operation will be modified to a minimal canonical form. Thus,
+`jj client-map set --from "" --to ""` will always completely wipe the map.
+The first matching rule in reverse list order determines how a particular
+folder/file should be named in the client.
+
+For simplicity, we require that all client-maps are _invertible_, that is:
```

Comment:
How do we require that? Do we check when the user updates the set that no two repo paths can possibly map to a single client path, or do we check only when we update the working copy? For example, is it valid to define patterns that map a->x and b->x/y, as long as there's no repo path starting with a/y (in the current working-copy commit), or is it an error regardless?